### PR TITLE
[clang-tidy] Add support for bsl::optional

### PR DESCRIPTION
**Pull request by ccotter**

### CODE REVIEW

**Review by PiotrZSL — APPROVED** [review]
I'm fine with change, but at the same time this check should be made configurable.
Also please update check documentation before committing.

**Comment by ccotter** [comment]
Agreed it should be configurable.

I updated the docs, would you mind merging if you are OK as I don't have perms?

**Review by ymand — CHANGES_REQUESTED** [review]

File: clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp [hunk 1]
```diff
@@ -38,10 +38,22 @@
 namespace clang {
 namespace dataflow {
 
-static bool isTopLevelNamespaceWithName(const NamespaceDecl &NS,
-                                        llvm::StringRef Name) {
-  return NS.getDeclName().isIdentifier() && NS.getName() == Name &&
-         NS.getParent() != nullptr && NS.getParent()->isTranslationUnit();
+template <class... NameTypes>
+static bool hasNestedNamespace(const NamespaceDecl &NS, llvm::StringRef Name,
```
[the remark this answers is not in the document]
Comment (ccotter) [reply clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp]: Good point, thanks! Updated.

**Comment by ccotter** [comment]
bump @ymand @PiotrZSL?

**Review by dayshah — APPROVED** [review]
huge for bloomberg 🔥

**Review by nicovank — APPROVED** [review]
LGTM. Rebase to fix release notes conflict and I'll merge 👍

**Pull request merged**